src/peft/tuners/vera/model.py

```diff
@@ -0,0 +1,501 @@
+
+    Returns:
+        `torch.nn.Module`: The Vera model.
+
+    Example:
+
+        ```py
+        >>> from transformers import AutoModelForSeq2SeqLM
+        >>> from peft import VeraModel, VeraConfig
+
+        >>> config = VeraConfig(
+        ...     task_type="SEQ_2_SEQ_LM",
+        ...     r=8,
+        ...     target_modules=["q", "v"],
+        ...     vera_dropout=0.01,
+        ... )
+
+        >>> model = AutoModelForSeq2SeqLM.from_pretrained("t5-base")
+        >>> vera_model = VeraModel(model, config)
+        ```
+
+        ```py
+        >>> import transformers
+        >>> from peft import VeraConfig, PeftModel, get_peft_model
+
+        >>> target_modules = ["q_proj", "k_proj", "v_proj", "out_proj", "fc_in", "fc_out", "wte"]
+        >>> config = VeraConfig(
+        ...     r=4, target_modules=target_modules, vera_dropout=0.1, bias="none", task_type="CAUSAL_LM"
+        ... )
+
+        >>> model = transformers.GPTJForCausalLM.from_pretrained(
+        ...     "kakaobrain/kogpt",
+        ...     revision="KoGPT6B-ryan1.5b-float16",  # or float32 version: revision=KoGPT6B-ryan1.5b
+        ...     pad_token_id=tokenizer.eos_token_id,
+        ...     use_cache=False,
+        ...     device_map={"": rank},
+        ...     torch_dtype=torch.float16,
+        ...     load_in_8bit=True,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> vera_model = get_peft_model(model, config)
+        ```
+
+    **Attributes**:
+        - **model** ([`~transformers.PreTrainedModel`]) -- The model to be adapted.
+        - **peft_config** ([`VeraConfig`]): The configuration of the Vera model.
+    """
+
+    prefix: str = "vera_lambda"
+
+    def __init__(self, model, config, adapter_name) -> None:
+        super().__init__(model, config, adapter_name)
+
+    def _find_first_dim(self, config) -> tuple[int, int]:
+        """
+        Finds the first linear layer that has been wrapped with Vera, and extract the input and output dimension.
+
+        This will be used for determining the size of the shared vera_A and vera_B matrices.
+
+        This will throw an error if there are multiple layers of the same type with different shapes.
+        """
+        model_config = getattr(self.model, "config", {"model_type": "custom"})
+        if hasattr(model_config, "to_dict"):
+            model_config = model_config.to_dict()
+
+        peft_config = self._prepare_adapter_config(config, model_config)
+        peft_config = _maybe_include_all_linear_layers(peft_config, self.model)
+
+        first_shape = None
+        for key, module in self.model.named_modules():
+            if not self._check_target_module_exists(peft_config, key):
+                continue
+
+            if isinstance(module, (nn.Linear, Conv1D)):
+                module_shape = tuple(module.weight.shape)
+                if isinstance(module, Conv1D):  # TODO: feels fragile, thoughts?
+                    module_shape = module_shape[::-1]
```

Comment:
remove this TODO? I feel this behaviour is actually fine, the semantics of Conv1D are unlikely to change

Reply:
Done.